supportedBackends/tes/src/main/scala/cromwell/backend/impl/tes/TesAsyncBackendJobExecutionActor.scala

```diff
@@ -59,6 +59,85 @@ case object Cancelled extends TesRunStatus {
+       |
+       |# Store token as environment variable
+       |export $environmentVariableName=$$(echo "$${sas_response_json}" | jq -r '.token')
+       |
+       |# Echo the first characters for logging/debugging purposes. "null" indicates something went wrong.
+       |echo Acquired sas token: "$${$environmentVariableName:0:4}****"
+       |### END ACQUIRE LOCAL SAS TOKEN ###
+       |""".stripMargin
+  }
+
+  private def maybeConvertToBlob(pathToTest: Try[Path]): Try[BlobPath] = {
+    pathToTest.collect { case blob: BlobPath => blob }
+  }
+
+  /**
+   * Under certain situations (and only on Terra), we want the VM running a TES task to have the ability to acquire a
+   * fresh sas token for itself. In order to be able to do this, we provide it with a precomputed endpoint it can use.
+   * The task VM will use the user assigned managed identity that it is running as in order to authenticate.
+   * We only return a value if at least one of the inputs is a BlobPath and //TODO flag specified in WDL.
+   * @param taskInputs The inputs to this particular TesTask. If any are blob files, the first  will be used to
+   *                   determine the storage container to retrieve the sas token for.
+   * @param pathGetter A function to convert string filepath into a cromwell Path object.
+   * @param blobConverter A function to convert a Path into a Blob path, if possible. Provided for testing purposes.
+   * @return A URL endpoint that, when called with proper authentication, will return a sas token.
+   *         Returns 'None' if one should not be used for this task.
+   */
+  def determineWSMSasEndpointFromInputs(taskInputs: List[Input],
+                                        pathGetter: String => Try[Path],
+                                        blobConverter: Try[Path] => Try[BlobPath] = maybeConvertToBlob): Option[String] = {
+    val shouldLocalizeSas = true //TODO: Make this a Workflow Option or come from the WDL
+    if (!shouldLocalizeSas) return None
+
+    // Collect all of the inputs that are valid blob paths
+    val blobFiles = taskInputs.collect{
+      case input if input.url.isDefined => BlobPathBuilder.validateBlobPath(input.url.get)
+    }.collect{
+      case valid: BlobPathBuilder.ValidBlobPath => valid
+    }
+    if(blobFiles.isEmpty) return None
+
+    // We use the first blob file in the list as a template for determining the localized sas params
+    val sasTokenEndpoint = for {
+      blob <- blobConverter(pathGetter(blobFiles.head.toUrl))
+      wsmEndpoint <- blob.wsmEndpoint
+      workspaceId <- blob.parseTerraWorkspaceIdFromPath
+      containerResourceId <- blob.containerWSMResourceId
+      endpoint = s"$wsmEndpoint/api/workspaces/v1/$workspaceId/resources/controlled/azure/storageContainer/$containerResourceId/getSasToken"
+    } yield endpoint
+
+    sasTokenEndpoint match {
+      case good: Success[String] => Some(good.value)
+      case _: Any => None
```

Comment:
This `Any` smells a little, why do we need to do this? Maybe the above `for` construct needs to be adjusted to ensure it has a `Try[String]` type? Also, more idiomatic to use:
```
case Success(good) => Option(good)
```